scheduler/core/components/optimizer/timeline.py

```diff
@@ -0,0 +1,62 @@
+EMPTY = -1
+UNSCHEDULABLE = -2
+
+from dataclasses import dataclass
+from typing import List, NoReturn, Mapping, Optional, Sized, Tuple
+from datetime import datetime, timedelta
+
+from lucupy.minimodel import Observation, ObservationID, Site
+from scheduler.core.calculations.nightevents import NightEvents
+
+import numpy as np
+
+
+@dataclass
+class Timeline:
+    """
+    Nightly plan for a specific Site for the GreedyMax optimizer. Each plan is a timeline array with one
+    entry for each time slot for the night. Each value needs to be a pointer or index to the observation
+    scheduled in that slot.
+    """
+    start: datetime
+    end: datetime
+    time_slot_length: timedelta
+    site: Site
+    _total_time_slots: int
+
+    def __post_init__(self):
+        self.schedule = np.full(self._total_time_slots, EMPTY)
+        self.is_full = False
+
+    def has(self, obs: Observation) -> bool:
+        return any(visit.obs_id == obs.id for visit in self.visits)
```

Comment:
A proposal here: instead of using a `has` method, I suggest we use the `__contains__` dunder method, i.e.:

```python
def __contains__(self, obs: Observation) -> bool:
    return any(visit.obs_id == obs.id for visit in self.visits)
```

Then instead of calling:
```python
timeline.has(obs)
```
or:
```python
not timeline.has(obs)
```

We can call instead:
```python
obs in timeline
```
or:
```python
obs not in timeline
```